plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java

```diff
@@ -174,134 +192,188 @@ public CompletableFuture<ConnectorSplitBatch> getNextBatch(int maxSize)
+        if (usedSimplifiedPredicate) {
+            // Pushed down predicate was simplified, always evaluate it against individual splits
+            this.pushedDownDynamicFilterPredicate = TupleDomain.all();
         }
 
-        TupleDomain<IcebergColumnHandle> dynamicFilterPredicate = dynamicFilter.getCurrentPredicate()
-                .transformKeys(IcebergColumnHandle.class::cast);
-        if (dynamicFilterPredicate.isNone()) {
+        TupleDomain<IcebergColumnHandle> effectivePredicate = dataColumnPredicate
+                .intersect(simplifiedPredicate);
+
+        if (effectivePredicate.isNone()) {
             finish();
-            return completedFuture(NO_MORE_SPLITS_BATCH);
+            return Stream.empty();
         }
 
-        List<ConnectorSplit> splits = new ArrayList<>(maxSize);
-        while (splits.size() < maxSize && (fileTasksIterator.hasNext() || fileScanIterator.hasNext())) {
-            if (!fileTasksIterator.hasNext()) {
-                FileScanTask wholeFileTask = fileScanIterator.next();
-                fileTasksIterator = wholeFileTask.split(targetSplitSize).iterator();
-                fileHasAnyDeletions = false;
-                // In theory, .split() could produce empty iterator, so let's evaluate the outer loop condition again.
-                continue;
-            }
-            FileScanTask scanTask = fileTasksIterator.next();
-            fileHasAnyDeletions = fileHasAnyDeletions || !scanTask.deletes().isEmpty();
-            if (scanTask.deletes().isEmpty() &&
-                    maxScannedFileSizeInBytes.isPresent() &&
-                    scanTask.file().fileSizeInBytes() > maxScannedFileSizeInBytes.get()) {
-                continue;
-            }
+        Expression filterExpression = toIcebergExpression(effectivePredicate);
+        // If the Dynamic Filter will be evaluated against each file, stats are required. Otherwise, skip them.
+        boolean requiresColumnStats = usedSimplifiedPredicate || !dynamicFilterIsComplete;
+        TableScan scan = tableScan.filter(filterExpression);
+        if (requiresColumnStats) {
+            scan = scan.includeColumnStats();
+        }
 
-            if (!pathDomain.includesNullableValue(utf8Slice(scanTask.file().path().toString()))) {
-                continue;
-            }
-            if (!fileModifiedTimeDomain.isAll()) {
-                long fileModifiedTime = getModificationTime(scanTask.file().path().toString());
-                if (!fileModifiedTimeDomain.includesNullableValue(packDateTimeWithZone(fileModifiedTime, UTC_KEY))) {
-                    continue;
-                }
-            }
-            IcebergSplit icebergSplit = toIcebergSplit(scanTask);
-
-            Schema fileSchema = scanTask.spec().schema();
-            Map<Integer, Optional<String>> partitionKeys = getPartitionKeys(scanTask);
-
-            Set<IcebergColumnHandle> identityPartitionColumns = partitionKeys.keySet().stream()
```

Comment:
the diff consist of new code being added and many lines moving around. It is hard for me to understand what are the important changes I should focus my review on.